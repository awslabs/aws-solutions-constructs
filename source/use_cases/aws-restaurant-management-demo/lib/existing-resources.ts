/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import * as cdk from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';

// Stack
// -
// This stack provisions an "existing resource" that simulates use of the restaurant management system with a piece of
// the restaurant's existing infrastructure. In this case, this stack provisions an order archive bucket to which
// the contents of the table in the use case example will output closed orders to.
// -
// In your environment, use the static method fromBucketName() to include an actual existing
// bucket from your environment in this stack. More info:
// https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-s3.Bucket.html#static-fromwbrbucketwbrnamescope-id-bucketname
// -
export class ExistingResources extends cdk.Stack {

  // Public variables
  public readonly archiveBucket: s3.Bucket;

  // Constructor
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // An "existing" Amazon S3 bucket that holds archived orders
    this.archiveBucket = new s3.Bucket(this, 'existing-order-archive-bucket');
  }
}
