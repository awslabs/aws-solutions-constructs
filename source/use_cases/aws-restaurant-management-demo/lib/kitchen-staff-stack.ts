/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as ddb from '@aws-cdk/aws-dynamodb';
import * as apigateway from '@aws-cdk/aws-apigateway';
import {
  CognitoToApiGatewayToLambda
} from '@aws-solutions-constructs/aws-cognito-apigateway-lambda';
import {
  LambdaToDynamoDB
} from '@aws-solutions-constructs/aws-lambda-dynamodb';

// Properties for the kitchen-staff-stack
export interface KitchenStaffStackProps {
  // The main database created in the shared-stack
  readonly db: ddb.Table,
}

// Stack
export class KitchenStaffStack extends cdk.Stack {

  // Constructor
  constructor(scope: cdk.Construct, id: string, props: KitchenStaffStackProps) {
    super(scope, id);

    // Create a Lambda function that lists all open orders in the database
    const getOpenOrders = new LambdaToDynamoDB(this, 'get-open-orders', {
      lambdaFunctionProps: {
        runtime: lambda.Runtime.NODEJS_14_X,
        code: lambda.Code.fromAsset(`${__dirname}/lambda/kitchen-staff/get-open-orders`),
        handler: 'index.handler',
        timeout: cdk.Duration.seconds(15)
      },
      existingTableObj: props.db
    });

    // Create a Lambda function that marks an order as completed in the database
    const completeOrder = new LambdaToDynamoDB(this, 'complete-order', {
      lambdaFunctionProps: {
        runtime: lambda.Runtime.NODEJS_14_X,
        code: lambda.Code.fromAsset(`${__dirname}/lambda/kitchen-staff/complete-order`),
        handler: 'index.handler',
        timeout: cdk.Duration.seconds(15)
      },
      existingTableObj: props.db
    });

    // Setup the kitchen staff API with Cognito user pool
    const kitchenStaffApi = new CognitoToApiGatewayToLambda(this, 'kitchen-staff-api', {
      existingLambdaObj: getOpenOrders.lambdaFunction,
      apiGatewayProps: {
        proxy: false,
        description: 'Demo: Kitchen staff API'
      }
    });

    // Add a resource to the API for listing all open orders
    const listOpenOrdersResource = kitchenStaffApi.apiGateway.root.addResource('get-open-orders');
    listOpenOrdersResource.addProxy({
      defaultIntegration: new apigateway.LambdaIntegration(kitchenStaffApi.lambdaFunction),
      anyMethod: true
    });

    // Add a resource to the API for completing orders
    const completeOrderResource = kitchenStaffApi.apiGateway.root.addResource('complete-order');
    completeOrderResource.addProxy({
      defaultIntegration: new apigateway.LambdaIntegration(completeOrder.lambdaFunction),
      anyMethod: true
    });

    // Add the authorizers to the API
    kitchenStaffApi.addAuthorizers();
  }
}