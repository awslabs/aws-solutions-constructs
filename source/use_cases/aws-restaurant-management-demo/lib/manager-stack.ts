/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as ddb from '@aws-cdk/aws-dynamodb';
import * as apigateway from '@aws-cdk/aws-apigateway';
import * as events from '@aws-cdk/aws-events';
import * as sfn from '@aws-cdk/aws-stepfunctions';
import * as sftasks from '@aws-cdk/aws-stepfunctions-tasks';
import * as s3 from '@aws-cdk/aws-s3';
import { CognitoToApiGatewayToLambda } from '@aws-solutions-constructs/aws-cognito-apigateway-lambda';
import { LambdaToDynamoDB } from '@aws-solutions-constructs/aws-lambda-dynamodb';
import { LambdaToS3 } from '@aws-solutions-constructs/aws-lambda-s3';
import { LambdaToSns } from '@aws-solutions-constructs/aws-lambda-sns';
import { EventsRuleToLambda } from '@aws-solutions-constructs/aws-events-rule-lambda';
import { LambdaToStepfunctions } from '@aws-solutions-constructs/aws-lambda-stepfunctions';

// Properties for the manager-stack
export interface ManagerStackProps {
  // The main database created in the shared-stack
  readonly db: ddb.Table,
  // The existing S3 bucket for orders to be archived to upon close-out
  readonly archiveBucket: s3.Bucket,
  // The Lambda layer for sharing database access functions
  readonly layer: lambda.LayerVersion
}

// Stack
export class ManagerStack extends cdk.Stack {

  // Constructor
  constructor(scope: cdk.Construct, id: string, props: ManagerStackProps) {
    super(scope, id);

    // Create a Lambda function that lists all orders from the database
    const getAllOrders = new LambdaToDynamoDB(this, 'get-all-orders', {
      lambdaFunctionProps: {
        runtime: lambda.Runtime.NODEJS_14_X,
        code: lambda.Code.fromAsset(`${__dirname}/lambda/manager/get-all-orders`),
        handler: 'index.handler',
        timeout: cdk.Duration.seconds(15),
        layers: [ props.layer ]
      },
      existingTableObj: props.db
    });

    // Create a Lambda function that will generate a report from information in the database
    // Runs as part of the close-out process
    const createReport = new LambdaToDynamoDB(this, 'create-report', {
      lambdaFunctionProps: {
        runtime: lambda.Runtime.NODEJS_14_X,
        code: lambda.Code.fromAsset(`${__dirname}/lambda/manager/create-report`),
        handler: 'index.handler',
        timeout: cdk.Duration.seconds(15),
        layers: [ props.layer ]
      },
      existingTableObj: props.db
    });

    // Create an S3 bucket for storing generated reports
    const reports = new LambdaToS3(this, 'reports-bucket', {
    	existingLambdaObj: createReport.lambdaFunction,
    });
    
    // Create a Lambda function that will calculate tips based on orders in the database
    // Runs as part of the close-out process
    const calculateTips = new LambdaToDynamoDB(this, 'calculate-tips', {
      lambdaFunctionProps: {
        runtime: lambda.Runtime.NODEJS_14_X,
        code: lambda.Code.fromAsset(`${__dirname}/lambda/manager/calculate-tips`),
        handler: 'index.handler',
        timeout: cdk.Duration.seconds(15),
        layers: [ props.layer ]
      },
      existingTableObj: props.db 
    });
    // Create a topic for tip reports to be sent to -> sends an email to service staff workers
    new LambdaToSns(this, 'calculate-tips-topic', {
      existingLambdaObj: calculateTips.lambdaFunction
    });

    // Create a Lambda function that will archive the items in the database to the existing S3 bucket
    // Runs as part of the close-out process
    const archiveOrders = new LambdaToDynamoDB(this, 'archive-orders', {
      lambdaFunctionProps: {
        runtime: lambda.Runtime.NODEJS_14_X,
        code: lambda.Code.fromAsset(`${__dirname}/lambda/manager/archive-orders`),
        handler: 'index.handler',
        timeout: cdk.Duration.seconds(15),
        layers: [ props.layer ]
      },
      existingTableObj: props.db
    });
    // Use the LambdaToS3 construct to connect the archiveOrders function with the existing bucket
    new LambdaToS3(this, 'archive-orders-to-bucket', {
      existingBucketObj: props.archiveBucket,
      existingLambdaObj: archiveOrders.lambdaFunction
    });

    // Setup the chain of events for the close-out process (via Step Function)
    // 1. Add a task for invoking the create-report function
    const createReportTask = new sfn.Task(this, 'create-reports-task', {
    	task: new sftasks.InvokeFunction(createReport.lambdaFunction)
    });
    // 2. Add a task for invoking the calculate-tips function
    const calculateTipsTask = new sfn.Task(this, 'calculate-tips-task', {
    	task: new sftasks.InvokeFunction(calculateTips.lambdaFunction)
    });
    // 3. Add a task for invoking the archive-orders function
    const archiveOrdersTask = new sfn.Task(this, 'archive-orders-task', {
    	task: new sftasks.InvokeFunction(archiveOrders.lambdaFunction)
    });
    // 4. Setup the chain
    const chain = sfn.Chain.start(createReportTask)
    	.next(calculateTipsTask)
    	.next(archiveOrdersTask);
    // 5. Setup the Step Functions integration with Lambda trigger
    const closeOutService = new LambdaToStepfunctions(this, 'close-out-service', {
      lambdaFunctionProps: {
        runtime: lambda.Runtime.NODEJS_14_X,
        code: lambda.Code.fromAsset(`${__dirname}/lambda/manager/close-out-service`),
        handler: 'index.handler',
        timeout: cdk.Duration.seconds(15)
      },
      stateMachineProps: {
    	  definition: chain
      }
    });

    // Create a Lambda function that will retrieve a specific report from the bucket
    const getReport = new LambdaToS3(this, 'get-report', {
      lambdaFunctionProps: {
        runtime: lambda.Runtime.NODEJS_14_X,
        code: lambda.Code.fromAsset(`${__dirname}/lambda/manager/get-report`),
        handler: 'index.handler',
        timeout: cdk.Duration.seconds(15)
      },
      existingBucketObj: reports.s3Bucket
    });
    
    // Setup the manager API with Cognito user pool
    const managerApi = new CognitoToApiGatewayToLambda(this, 'manager-api', {
    	existingLambdaObj: getAllOrders.lambdaFunction,
		  apiGatewayProps: {
	      proxy: false,
        description: 'Demo: Manager API'
	    }
    });
    
    // Add a resource to the API for listing all orders
    const listOrdersResource = managerApi.apiGateway.root.addResource('get-all-orders');
    listOrdersResource.addProxy({
    	defaultIntegration: new apigateway.LambdaIntegration(managerApi.lambdaFunction),
    	anyMethod: true
    });
    
    // Add a resource to the API for triggering the close-out process
    const closeOutServiceResource = managerApi.apiGateway.root.addResource('close-out-service');
    closeOutServiceResource.addProxy({
    	defaultIntegration: new apigateway.LambdaIntegration(closeOutService.lambdaFunction),
    	anyMethod: true
    });
    
    // Add a resource to the API for viewing reports
    const getReportResource = managerApi.apiGateway.root.addResource('get-report');
    getReportResource.addProxy({
    	defaultIntegration: new apigateway.LambdaIntegration(getReport.lambdaFunction),
    	anyMethod: true
    });
    
    // Add the authorizers to the API
    managerApi.addAuthorizers();
    
    // Create a Lambda function for identifying orders that have been open for too long
    const checkLateOrders = new LambdaToDynamoDB(this, 'check-late-orders', {
      lambdaFunctionProps: {
        runtime: lambda.Runtime.NODEJS_14_X,
        code: lambda.Code.fromAsset(`${__dirname}/lambda/manager/check-late-orders`),
        handler: 'index.handler',
        environment: {
          LATE_ORDER_THRESHOLD: '30'
        },
        timeout: cdk.Duration.seconds(15)
      },
      existingTableObj: props.db
    });
    
    // Create a CloudWatch Events rule to check for late orders every minute
    new EventsRuleToLambda(this, 'check-late-orders-scheduler', {
    	existingLambdaObj: checkLateOrders.lambdaFunction,
    	eventRuleProps: {
	      schedule: events.Schedule.rate(cdk.Duration.minutes(1))
	    }
    });
    
    // Create an SNS topic to send notifications to the manager when one or more orders are late
    new LambdaToSns(this, 'check-late-orders-notifier', {
    	existingLambdaObj: checkLateOrders.lambdaFunction
    });
  }
}