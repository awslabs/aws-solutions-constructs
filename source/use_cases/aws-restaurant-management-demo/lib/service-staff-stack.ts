/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as ddb from '@aws-cdk/aws-dynamodb';
import * as apigateway from '@aws-cdk/aws-apigateway';
import {
  CognitoToApiGatewayToLambda
} from '@aws-solutions-constructs/aws-cognito-apigateway-lambda';
import {
  LambdaToDynamoDB
} from '@aws-solutions-constructs/aws-lambda-dynamodb';

// Properties for the service-staff-stack
export interface ServiceStaffStackProps {
  // The main database created in the shared-stack
  readonly db: ddb.Table,
}

// Stack
export class ServiceStaffStack extends cdk.Stack {
  
  // Constructor
  constructor(scope: cdk.Construct, id: string, props: ServiceStaffStackProps) {
    super(scope, id);

    // Create a Lambda function that adds a new order to the database
    const createOrder = new LambdaToDynamoDB(this, 'create-order', {
      lambdaFunctionProps: {
        runtime: lambda.Runtime.NODEJS_14_X,
        code: lambda.Code.fromAsset(`${__dirname}/lambda/service-staff/create-order`),
        handler: 'index.handler',
        timeout: cdk.Duration.seconds(15)
      },
      existingTableObj: props.db
    });

    // Create a Lambda function that closes out an order in the table
    const processPayment = new LambdaToDynamoDB(this, 'process-payment', {
      lambdaFunctionProps: {
        runtime: lambda.Runtime.NODEJS_14_X,
        code: lambda.Code.fromAsset(`${__dirname}/lambda/service-staff/process-payment`),
        handler: 'index.handler',
        timeout: cdk.Duration.seconds(15)
      },
      existingTableObj: props.db
    });

    // Setup the service staff API with Cognito user pool
    const serviceStaffApi = new CognitoToApiGatewayToLambda(this, 'service-staff-api', {
    	existingLambdaObj: createOrder.lambdaFunction,
		  apiGatewayProps: {
	      proxy: false,
        description: 'Demo: Service staff API'
	    }
    });
    
    // Add a resource to the API for creating a new order
    const createOrderResource = serviceStaffApi.apiGateway.root.addResource('create-order');
    createOrderResource.addProxy({
    	defaultIntegration: new apigateway.LambdaIntegration(serviceStaffApi.lambdaFunction),
    	anyMethod: true
    });
    
    // Add a resource to the API for handling payments and marking orders as paid
    const processPaymentResource = serviceStaffApi.apiGateway.root.addResource('process-payment');
    processPaymentResource.addProxy({
    	defaultIntegration: new apigateway.LambdaIntegration(processPayment.lambdaFunction),
    	anyMethod: true
    });
    
    // Add the authorizers to the API
    serviceStaffApi.addAuthorizers();
  }
}