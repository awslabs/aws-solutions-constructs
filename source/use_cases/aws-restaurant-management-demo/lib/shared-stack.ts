/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import * as cdk from '@aws-cdk/core';
import * as ddb from '@aws-cdk/aws-dynamodb';
import * as lambda from '@aws-cdk/aws-lambda';

// Stack
export class SharedStack extends cdk.Stack {

  // Public variables
  public readonly database: ddb.Table;
  public readonly layer: lambda.LayerVersion;

  // Constructor
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Setup the database ----------------------------------------------------------------------------------------------
    this.database = new ddb.Table(this, "order-table", {
      partitionKey: {
        name: "id",
        type: ddb.AttributeType.STRING,
      },
      billingMode: ddb.BillingMode.PROVISIONED,
      removalPolicy: cdk.RemovalPolicy.DESTROY
    });

    // Add autoscaling
    const readScaling = this.database.autoScaleReadCapacity({
      minCapacity: 1,
      maxCapacity: 50,
    });

    readScaling.scaleOnUtilization({
      targetUtilizationPercent: 50,
    });

    // Add a global secondary index for query operations
    this.database.addGlobalSecondaryIndex({
      partitionKey: {
        name: "gsi1pk",
        type: ddb.AttributeType.STRING,
      },
      sortKey: {
        name: "gsi1sk",
        type: ddb.AttributeType.STRING,
      },
      indexName: "gsi1pk-gsi1sk-index",
    });

    // Setup a Lambda layer for sharing database functions -------------------------------------------------------------
    this.layer = new lambda.LayerVersion(this, 'shared-db-functions-layer', {
      code: lambda.Code.fromAsset(`${__dirname}/lambda/layer`),
      compatibleRuntimes: [ lambda.Runtime.NODEJS_14_X ],
      license: 'Apache-2.0',
      description: 'Layer for common database access functions',
    });
  }
}