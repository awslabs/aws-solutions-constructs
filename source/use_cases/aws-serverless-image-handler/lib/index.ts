/**
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import * as defaults from '@aws-solutions-constructs/core';
import { Construct } from '@aws-cdk/core';
import * as cloudFront from '@aws-cdk/aws-cloudfront';
import * as apiGateway from '@aws-cdk/aws-apigateway';
import * as lambda from '@aws-cdk/aws-lambda';
import * as iam from '@aws-cdk/aws-iam';
import * as s3 from '@aws-cdk/aws-s3';
import { CloudFrontToApiGatewayToLambda } from '@aws-solutions-constructs/aws-cloudfront-apigateway-lambda';
import { LambdaToS3 } from '@aws-solutions-constructs/aws-lambda-s3';

/**
 * The properties for the ServerlessImageHandler class.
 */
export interface ServerlessImageHandlerProps {
    /**
     * Whether or not to emable Cross-Origin Resource Sharing (CORS) for the image handler API.
     *
     * @default - false.
     */
    readonly corsEnabled?: boolean
    /**
     * The CORS origin to use for the image handler API. This property is only required if `corsEnabled` is set to true.
     *
     * @default - none.
     */
    readonly corsOrigin?: string
    /**
     * One or more buckets within the deployment account to be used for storing/sourcing original image files.
     *
     * @default - required.
     */
    readonly sourceBuckets: string
    /**
     * The amount of time for CloudWatch log entries from this solution to be retained.
     *
     * @default - 1 day.
     */
    readonly logRetentionPeriod?: number,
    /**
     * Whether or not to accept/enable automatic WebP based on Accept- headers.
     *
     * @default - false.
     */
    readonly autoWebP?: boolean,
    /**
     * Optional user provided props to override the default props for each resource.
     *
     * @default - undefined/optional.
     */
    readonly customProps?: ServerlessImageHandlerCustomProps
}

/**
 * Custom properties for the ServerlessImageHandler class.
 */
export interface ServerlessImageHandlerCustomProps {
    /**
     * Optional user provided props to override the default props for the CloudFront distribution.
     *
     * @default - false.
     */
    readonly cloudFrontDistributionProps?: cloudFront.CloudFrontWebDistributionProps | any,
    /**
     * Optional user provided props to override the default props for the API Gateway REST API.
     *
     * @default - none.
     */
    readonly apiGatewayProps?: apiGateway.RestApiProps | any,
    /**
     * Optional user provided props to override the default props for the Lambda function.
     *
     * @default - none.
     */
    readonly lambdaFunctionProps?: lambda.FunctionProps | any,
    /**
     * Optional user provided props to override the default props for the S3 bucket.
     *
     * @default - none.
     */
    readonly bucketProps?: s3.BucketProps | any,
    /**
     * Optional user provided props to override the default permissions for the S3 bucket.
     *
     * @default - none.
     */
    readonly bucketPermissions?: string[]
}

/**
 * @summary The ServerlessImageHandler class.
 */
export class ServerlessImageHandler extends Construct {
    // Private variables
    private cloudFrontApiGatewayLambda: CloudFrontToApiGatewayToLambda;
    private lambdaS3: LambdaToS3;
    private customProps: any;

    /**
     * @summary Constructs a new instance of the ServerlessImageHandler class.
     * @param {cdk.App} scope - represents the scope for all the resources.
     * @param {string} id - this is a a scope-unique id.
     * @param {ServerlessImageHandlerProps} props - user provided props for the construct
     * @since 0.8.0
     * @access public
     */
    constructor(scope: Construct, id: string, props: ServerlessImageHandlerProps) {
        super(scope, id);

        // If customProps is undefined, define it
        this.customProps = (props.customProps === undefined) ? {} : props.customProps;

        // Use case specific properties for the Lambda function
        const useCaseFunctionProps: lambda.FunctionProps = {
            code: lambda.Code.fromAsset(`${__dirname}/lambda/image-handler`),
            runtime: lambda.Runtime.NODEJS_12_X,
            handler: 'index.handler',
            environment: {
                AUTO_WEBP: (props.autoWebP) ? 'Yes' : 'No',
                CORS_ENABLED: (props.corsEnabled) ? 'Yes' : 'No',
                CORS_ORIGIN: (props.corsOrigin) ? props.corsOrigin : ''
            }
        };
        const functionProps = (this.customProps.lambdaFunctionProps) ?
            defaults.overrideProps(useCaseFunctionProps, this.customProps.lambdaFunctionProps) : useCaseFunctionProps;

        // Use case specific properties for the API Gateway
        const useCaseApiProps: apiGateway.RestApiProps = {
            binaryMediaTypes: [ "*/*" ]
        };
        const apiProps = (this.customProps.apiGatewayProps) ?
            defaults.overrideProps(useCaseApiProps, this.customProps.apiGatewayProps) : useCaseApiProps;

        // Build the CloudFrontToApiGatewayToLambda pattern
        this.cloudFrontApiGatewayLambda = new CloudFrontToApiGatewayToLambda(this, 'CloudFrontApiGatewayLambda', {
            cloudFrontDistributionProps: (this.customProps.cloudFrontDistributionProps) ? this.customProps.cloudFrontDistributionProps : undefined,
            apiGatewayProps: apiProps,
            lambdaFunctionProps: functionProps
        });
        const existingLambdaFn = this.cloudFrontApiGatewayLambda.lambdaFunction;

        // Build the LambdaToS3 pattern
        this.lambdaS3 = new LambdaToS3(this, 'ExistingLambdaS3', {
            existingLambdaObj: existingLambdaFn,
            bucketProps: this.customProps.bucketProps,
            bucketPermissions: (this.customProps.bucketPermissions) ? this.customProps.bucketPermissions : undefined
        });

        // Add additional permissions for Lambda to source original images from any bucket in the account
        const lambdaSourcingPolicyStmt = new iam.PolicyStatement({
            effect: iam.Effect.ALLOW
        });
        lambdaSourcingPolicyStmt.addResources('arn:aws:s3:::*');
        lambdaSourcingPolicyStmt.addActions('s3:GetObject*', 's3:GetBucket*', 's3:List*');

        // Add additional permissions for Lambda to access Rekognition services
        const lambdaRekognitionPolicyStmt = new iam.PolicyStatement({
            effect: iam.Effect.ALLOW
        });
        lambdaRekognitionPolicyStmt.addResources('*');
        lambdaRekognitionPolicyStmt.addActions('rekognition:DetectFaces');

        // Append the additional permissions to an inline policy
        const inlinePolicy = new iam.Policy(this, 'LambdaS3AccessPolicy', {
            statements: [ lambdaSourcingPolicyStmt, lambdaRekognitionPolicyStmt ]
        });

        // Add cfn_nag suppression for Rekognition wildcard resource
        const rawInlinePolicy: iam.CfnPolicy = inlinePolicy.node.findChild('Resource') as iam.CfnPolicy;
        rawInlinePolicy.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W12',
                    reason: `Specified Rekognition action needs wildcard resource.`
                }]
            }
        };

        // Attach the inline policy to the Lambda function role
        existingLambdaFn.role?.attachInlinePolicy(inlinePolicy);

        // Add the SOURCE_BUCKETS environment variable to the Lambda function
        const bucketsArr = (props.sourceBuckets !== "") ? props.sourceBuckets.split(',') : [];
        bucketsArr.push(this.lambdaS3.s3Bucket.bucketName);
        const bucketsStr = bucketsArr.toString().replace(/\s+/g, '');
        this.cloudFrontApiGatewayLambda.lambdaFunction.addEnvironment("SOURCE_BUCKETS", bucketsStr);
    }

    /**
     * @summary Returns an instance of cloudFront.CloudFrontWebDistribution created by the construct.
     * @returns { cloudFront.CloudFrontWebDistribution } Instance of CloudFrontWebDistribution created by the construct.
     * @since 0.8.0
     * @access public
     */
    public cloudFrontDistribution(): cloudFront.CloudFrontWebDistribution {
        return this.cloudFrontApiGatewayLambda.cloudFrontWebDistribution;
    }

    /**
     * @summary Returns an instance of apiGateway.RestApi created by the construct.
     * @returns { apiGateway.RestApi } Instance of RestApi created by the construct.
     * @since 0.8.0
     * @access public
     */
    public apiGateway(): apiGateway.RestApi {
        return this.cloudFrontApiGatewayLambda.apiGateway;
    }

    /**
     * @summary Returns an instance of lambda.Function created by the construct.
     * @returns { lambda.Function } Instance of Function created by the construct
     * @since 0.8.0
     * @access public
     */
    public lambdaFunction(): lambda.Function {
        return this.cloudFrontApiGatewayLambda.lambdaFunction;
    }

    /**
     * @summary Returns an instance of s3.Bucket created by the construct.
     * @returns { s3.Bucket } Instance of Bucket created by the construct
     * @since 0.8.0
     * @access public
     */
    public s3Bucket(): s3.Bucket {
        return this.lambdaS3.s3Bucket;
    }
}