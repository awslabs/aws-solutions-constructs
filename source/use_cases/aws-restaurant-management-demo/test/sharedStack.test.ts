/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

 import * as cdk from '@aws-cdk/core';
 import { SharedStack } from '../lib/shared-stack';
 import { SynthUtils } from '@aws-cdk/assert';
 import '@aws-cdk/assert/jest';
 
 // Environment configuration
 const config = { 
   env: {
     account: 'ACCOUNT_NUMBER_HERE', 
     region: 'us-east-1' // default region selection
   }
 };
 
 test('test-shared-stack', () => {
   const app = new cdk.App();
   const stack = new SharedStack(app, `SharedStack`, config);
   expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
 });