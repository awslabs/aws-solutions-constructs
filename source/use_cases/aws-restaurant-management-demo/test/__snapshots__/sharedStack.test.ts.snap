// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-shared-stack 1`] = `
Object {
  "Resources": Object {
    "databaseEBDE4557": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "orderStatus",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "timeOpened",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "username",
            "AttributeType": "S",
          },
        ],
        "GlobalSecondaryIndexes": Array [
          Object {
            "IndexName": "orderStatus_index",
            "KeySchema": Array [
              Object {
                "AttributeName": "id",
                "KeyType": "HASH",
              },
              Object {
                "AttributeName": "orderStatus",
                "KeyType": "RANGE",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
            "ProvisionedThroughput": Object {
              "ReadCapacityUnits": 5,
              "WriteCapacityUnits": 5,
            },
          },
          Object {
            "IndexName": "timeOpened_index",
            "KeySchema": Array [
              Object {
                "AttributeName": "id",
                "KeyType": "HASH",
              },
              Object {
                "AttributeName": "timeOpened",
                "KeyType": "RANGE",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
            "ProvisionedThroughput": Object {
              "ReadCapacityUnits": 5,
              "WriteCapacityUnits": 5,
            },
          },
          Object {
            "IndexName": "username_index",
            "KeySchema": Array [
              Object {
                "AttributeName": "id",
                "KeyType": "HASH",
              },
              Object {
                "AttributeName": "username",
                "KeyType": "RANGE",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
            "ProvisionedThroughput": Object {
              "ReadCapacityUnits": 5,
              "WriteCapacityUnits": 5,
            },
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
